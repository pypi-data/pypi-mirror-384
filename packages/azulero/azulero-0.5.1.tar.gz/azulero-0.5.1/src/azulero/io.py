# SPDX-FileCopyrightText: Copyright (C) 2025, Antoine Basset
# SPDX-PackageSourceInfo: https://github.com/kabasset/azulero
# SPDX-License-Identifier: Apache-2.0

from astropy.io import fits
import numpy as np
import tifffile
from pathlib import Path


def parse_tile(text: str):
    tile_slicing = text.split("[")
    if len(tile_slicing) == 1:
        return tile_slicing[0], None
    return tile_slicing[0], parse_slice(tile_slicing[1][:-1])


def parse_slice(text: str):
    """
    Parse a 2D slice, e.g. ":,3:14".
    """
    if text is None:
        return None
    parse_index = lambda i: int(i) if i else None
    return tuple(
        slice(*[parse_index(i) for i in axis.split(":")]) for axis in text.split(",")
    )


def parse_map(text: str, dtype=float):
    """
    Parse a comma-separated list of 'key:value' pairs.
    """
    if not text:
        return []
    pairs = [p.split(":") for p in text.split(",")]
    return [[dtype(x), dtype(y)] for x, y in pairs]


def read_fits(path: Path, slicing=None):
    """
    Read a region in the primary array of a FITS file.
    """
    data = fits.getdata(path)
    return data if slicing is None else data[slicing]


def make_workdir(workspace, tile):
    workdir = Path(workspace).expanduser() / tile
    if workdir.is_dir():
        print("WARNING: Working directory already exists.")
    else:
        workdir.mkdir(parents=True)
    return workdir


def write_fits(data: np.array, path: Path):
    """
    Write an SIF file.
    """
    fits.PrimaryHDU(data).writeto(path, overwrite=True)


def write_tiff(rgb: np.ndarray, path: Path):
    """
    Write a normalized RGB image.
    """
    data = np.flipud(rgb * 65535).astype(np.uint16)
    tifffile.imwrite(path, data)


def read_channel(workdir: Path, channel: str, slicing=None):
    """
    Read the region of one channel.
    """
    data_files = list(workdir.glob(f"EUC_*{channel}_*.fits"))
    assert len(data_files) == 1
    return read_fits(data_files[0], slicing)


def read_iyjh(workdir: Path, slicing=None):
    """
    Read the region of a VIS- and NIR-covered tile.
    """
    return np.stack(
        (
            read_channel(workdir, "VIS", slicing),
            read_channel(workdir, "NIR-Y", slicing),
            read_channel(workdir, "NIR-J", slicing),
            read_channel(workdir, "NIR-H", slicing),
        )
    )
