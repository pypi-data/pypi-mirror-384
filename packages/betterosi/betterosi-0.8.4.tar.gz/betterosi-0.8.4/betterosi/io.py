import struct
from pathlib import Path
from typing import Any

from mcap_protobuf.decoder import DecoderFactory
from mcap.reader import make_reader
from mcap_protobuf.writer import Writer as McapWriter
from .generated import osi3

MESSAGES_TYPE = [
    "SensorView",
    "SensorViewConfiguration",
    "GroundTruth",
    "HostVehicleData",
    "SensorData",
    "TrafficCommand",
    "TrafficCommandUpdate",
    "TrafficUpdate",
    "MotionRequest",
    "StreamingUpdate",
    "MapAsamOpenDrive",
]


def gen2betterosi(
    schema,
    message,
    return_sensor_view=False,
    return_ground_truth=False,
    passthrough=False,
):
    if not passthrough:
        if any(schema.name == f"osi3.{k}" for k in MESSAGES_TYPE):
            message_cls = getattr(osi3, schema.name.split(".")[-1])
            message = message_cls().parse(message.SerializeToString())
        else:
            return None
    if not return_sensor_view and not return_ground_truth:
        return message
    if return_sensor_view and schema.name == "osi3.SensorView":
        return message
    if return_ground_truth:
        if schema.name == "osi3.SensorView":
            return message.global_ground_truth
        if schema.name == "osi3.GroundTruth":
            return message
    return None


def iter_osi_trace_file(f, m):
    while True:
        length_bytes = f.read(4)
        if not length_bytes:
            break  # EOF
        if len(length_bytes) < 4:
            raise ValueError("Truncated length header")
        (msg_len,) = struct.unpack("<I", length_bytes)
        message = f.read(msg_len)
        if len(message) < msg_len:
            raise ValueError("Truncated message body")
        yield m.parse(message)


def read(
    filepath: str,
    return_sensor_view=False,
    return_ground_truth=False,
    mcap_return_betterosi: bool = True,
    mcap_topics: list | None = None,
    osi_message_type: str | None = None,
) -> list[Any]:
    p = Path(filepath)
    with p.open("rb") as f:
        if p.suffix == ".mcap":
            reader = make_reader(f, decoder_factories=[DecoderFactory()])
            views = (
                gen2betterosi(
                    schema,
                    proto_msg,
                    return_sensor_view=return_sensor_view,
                    return_ground_truth=return_ground_truth,
                    passthrough=not mcap_return_betterosi,
                )
                for schema, channel, message, proto_msg in reader.iter_decoded_messages(
                    topics=mcap_topics
                )
            )
            views = (v for v in views if v is not None)
        elif p.suffix == ".osi":
            if return_sensor_view or return_ground_truth:
                try:
                    with p.open("r") as t:
                        _ = next(iter_osi_trace_file(t, osi3.SensorView))
                        is_sv = True
                except Exception as e:
                    if return_sensor_view:
                        raise e
                    is_sv = False
                if is_sv:
                    views = iter_osi_trace_file(f, osi3.SensorView)
                    if not return_sensor_view:
                        views = (m.global_ground_truth for m in views)
                else:
                    views = iter_osi_trace_file(f, osi3.GroundTruth)
            else:
                if osi_message_type is None:
                    raise ValueError(
                        "Specify the osi_message_type, e.g., `GroundTruth`."
                    )
                views = iter_osi_trace_file(f, getattr(osi3, osi_message_type))
        else:
            raise NotImplementedError()
        for v in views:
            yield v


class Writer:
    def __init__(self, output, topic="ground_truth", mode="wb", **kwargs):
        p = Path(output)
        if p.suffix == ".mcap":
            self.write_mcap = True
            self.write_osi = False
            self.topic = topic
            self.file = open(p, mode)
            self.mcap_writer = McapWriter(self.file, **kwargs)
        elif p.suffix == ".osi":
            self.write_mcap = False
            self.write_osi = True
            self.file = open(p, mode)
        else:
            raise NotImplementedError()

    def __enter__(self):
        return self

    def add(self, view, topic: str = None, log_time=None):
        if self.write_mcap:
            if log_time is None:
                log_time = int(view.timestamp.nanos + view.timestamp.seconds * 1e9)
            topic = self.topic if topic is None else topic
            (
                self.mcap_writer.write_message(
                    topic, view, log_time=log_time, publish_time=log_time
                ),
            )
        if self.write_osi:
            buffer = bytes(view)
            self.file.write(struct.pack("<L", len(buffer)))
            self.file.write(buffer)

    def __exit__(self, exc_: Any, exc_type_: Any, tb_: Any):
        if self.write_mcap:
            self.mcap_writer.finish()
        self.file.close()
