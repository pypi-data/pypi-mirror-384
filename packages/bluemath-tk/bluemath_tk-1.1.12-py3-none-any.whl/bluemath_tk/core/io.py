import pickle

from .models import BlueMathModel


def load_model(model_path: str) -> BlueMathModel:
    """
    Loads a BlueMathModel from a file.

    Parameters
    ----------
    model_path : str
        The path to the model file.

    Returns
    -------
    BlueMathModel
        The loaded BlueMathModel.
    """

    with open(model_path, "rb") as f:
        model = pickle.load(f)

    return model
