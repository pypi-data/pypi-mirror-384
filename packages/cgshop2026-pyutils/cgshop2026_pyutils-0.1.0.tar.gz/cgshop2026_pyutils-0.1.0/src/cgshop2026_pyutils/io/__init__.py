import functools
from pathlib import Path

from ..schemas.instance import CGSHOP2026Instance
from ..schemas.solution import CGSHOP2026Solution


def open_file(func):
    """
    Decorator to open a file before calling the function and close it afterwards,
    if passed as string or pathlib.Path.
    """

    @functools.wraps(func)
    def wrapper(file, *args, **kwargs):
        if isinstance(file, str):
            file = Path(file)
        if isinstance(file, Path):
            with file.open() as f:
                return func(f, *args, **kwargs)
        return func(file, *args, **kwargs)

    return wrapper


@open_file
def read_instance(file) -> CGSHOP2026Instance:
    """
    Read an instance from a file.
    :param file: File object or path to the file.
    :return: Instance object
    """
    content = file.read()
    return CGSHOP2026Instance.model_validate_json(content)


@open_file
def read_solution(file) -> CGSHOP2026Solution:
    """
    Read a solution from a file.
    :param file: File object or path to the file.
    :return: Solution object
    """
    content = file.read()
    return CGSHOP2026Solution.model_validate_json(content)
