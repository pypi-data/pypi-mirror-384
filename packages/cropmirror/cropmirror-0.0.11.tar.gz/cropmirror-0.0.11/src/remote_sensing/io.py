"""Input/Output operations for GeoTIFF and raster data."""
from typing import Tuple, Optional
import numpy as np
from osgeo import gdal, ogr, osr
import logging


def read_tiff(imgfile: str) -> Tuple[np.ndarray, int, int, tuple, str]:
    """Read GeoTIFF file and return data and metadata.
    
    Args:
        imgfile: Path to GeoTIFF file.
        
    Returns:
        Tuple of (dataset, width, height, geotransform, projection).
    """
    data = gdal.Open(imgfile)
    if data is None:
        raise FileNotFoundError(f"Cannot open file: {imgfile}")
    
    dataset = data.ReadAsArray()
    im_width = data.RasterXSize
    im_height = data.RasterYSize
    im_geotrans = data.GetGeoTransform()
    im_proj = data.GetProjection()
    
    data = None  # Close file
    return dataset, im_width, im_height, im_geotrans, im_proj


def save_tiff(
    savename: str,
    saveshp: str,
    im_width: int,
    im_height: int,
    im_geotrans: tuple,
    im_proj: str,
    datai: np.ndarray,
    mask: np.ndarray,
    do_clustering: bool = True,
    apply_mask_2: bool = True
) -> None:
    """Save processed data to GeoTIFF and shapefile.
    
    Args:
        savename: Output GeoTIFF file path.
        saveshp: Output shapefile path.
        im_width: Image width.
        im_height: Image height.
        im_geotrans: GeoTransform tuple.
        im_proj: Projection string.
        datai: Data array to save.
        mask: Mask array.
        do_clustering: Whether clustering was performed.
        apply_mask_2: Whether to apply mask==2 filter.
    """
    logging.info(f"Saving GeoTIFF to {savename}")
    
    # Check if clustering is needed
    valid_data = datai[~np.isnan(datai)]
    unique_values = np.unique(valid_data)
    should_cluster = do_clustering and len(unique_values) > 1
    
    cluster_centers = None
    label_to_center = {}
    
    # Perform clustering if needed
    if should_cluster:
        logging.info("Performing clustering for output")
        from sklearn.cluster import KMeans
        
        kmeans = KMeans(n_clusters=min(5, len(unique_values)), random_state=42)
        valid_data_reshaped = datai[~np.isnan(datai)].reshape(-1, 1)
        
        if len(valid_data_reshaped) > 0:
            kmeans.fit(valid_data_reshaped)
            labels = kmeans.labels_
            cluster_centers = kmeans.cluster_centers_.flatten()
            
            # Map cluster labels to data
            clustered_data = np.full_like(datai, np.nan)
            clustered_data[~np.isnan(datai)] = labels + 1
            datai = clustered_data
            
            # Create label to center mapping
            label_to_center = {label + 1: center for label, center in enumerate(cluster_centers)}
    else:
        logging.info("Skipping clustering")
        label_to_center = {1: np.mean(valid_data)} if len(valid_data) > 0 else {1: 0}
    
    # Apply mask filter
    if apply_mask_2:
        datai = np.where(mask == 2, datai, np.nan)
    else:
        datai = np.where(mask != 0, datai, np.nan)
    
    # Save GeoTIFF
    driver = gdal.GetDriverByName("GTiff")
    dataset = driver.Create(savename, im_width, im_height, 1, gdal.GDT_Float32)
    dataset.SetGeoTransform(im_geotrans)
    dataset.SetProjection(im_proj)
    dataset.GetRasterBand(1).SetNoDataValue(np.nan)
    dataset.GetRasterBand(1).WriteArray(datai)
    
    # Create shapefile
    _create_shapefile(
        saveshp, im_proj, dataset, datai, 
        im_width, im_height, im_geotrans, label_to_center
    )
    
    dataset = None  # Close file
    logging.info(f"Successfully saved {savename}")


def _create_shapefile(
    saveshp: str,
    im_proj: str,
    dataset: gdal.Dataset,
    datai: np.ndarray,
    im_width: int,
    im_height: int,
    im_geotrans: tuple,
    label_to_center: dict
) -> None:
    """Create shapefile from raster data.
    
    Args:
        saveshp: Output shapefile path.
        im_proj: Projection string.
        dataset: GDAL dataset.
        datai: Data array.
        im_width: Image width.
        im_height: Image height.
        im_geotrans: GeoTransform tuple.
        label_to_center: Mapping from labels to cluster centers.
    """
    import os
    
    drv = ogr.GetDriverByName("ESRI Shapefile")
    if os.path.exists(saveshp):
        drv.DeleteDataSource(saveshp)
    
    polygon = drv.CreateDataSource(saveshp)
    proj_raster = osr.SpatialReference()
    proj_raster.ImportFromWkt(im_proj)
    poly_layer = polygon.CreateLayer("shp", srs=proj_raster, geom_type=ogr.wkbMultiPolygon)
    
    # Add value field
    value_field = ogr.FieldDefn("value", ogr.OFTReal)
    poly_layer.CreateField(value_field)
    
    # Create mask for valid data
    mask_array = np.where(~np.isnan(datai), 1, 0)
    driver = gdal.GetDriverByName("GTiff")
    mask_dataset = driver.Create('/vsimem/mask.tif', im_width, im_height, 1, gdal.GDT_Byte)
    mask_dataset.SetGeoTransform(im_geotrans)
    mask_dataset.SetProjection(im_proj)
    mask_dataset.GetRasterBand(1).WriteArray(mask_array)
    
    # Polygonize
    gdal.FPolygonize(dataset.GetRasterBand(1), mask_dataset.GetRasterBand(1), poly_layer, 0)
    
    # Update feature values with cluster centers
    for feature in poly_layer:
        label = feature.GetField(0)
        if label is not None and label in label_to_center:
            center_value = label_to_center[label]
            feature.SetField("value", float(center_value))
            poly_layer.SetFeature(feature)
    
    # Cleanup
    mask_dataset = None
    gdal.Unlink('/vsimem/mask.tif')
    polygon.SyncToDisk()
    polygon = None


def read_img(filename: str) -> Tuple[str, tuple, np.ndarray, int, int, int]:
    """Read image file with GDAL.
    
    Args:
        filename: Path to image file.
        
    Returns:
        Tuple of (projection, geotransform, data, width, height, band_count).
    """
    dataset = gdal.Open(filename)
    if dataset is None:
        raise FileNotFoundError(f"Cannot find/open {filename}")
    
    im_width = dataset.RasterXSize
    im_height = dataset.RasterYSize
    im_band = dataset.RasterCount
    im_geotrans = dataset.GetGeoTransform()
    im_proj = dataset.GetProjection()
    im_data = dataset.ReadAsArray(0, 0, im_width, im_height)
    
    dataset = None
    return im_proj, im_geotrans, im_data, im_width, im_height, im_band


def write_img(filename: str, im_proj: str, im_geotrans: tuple, im_data: np.ndarray) -> None:
    """Write image to file with GDAL.
    
    Args:
        filename: Output file path.
        im_proj: Projection string.
        im_geotrans: GeoTransform tuple.
        im_data: Image data array.
    """
    import pathlib
    
    # Determine data type
    dtype_to_gdal = {
        "uint8": gdal.GDT_Byte,
        "uint16": gdal.GDT_UInt16,
        "int16": gdal.GDT_Int16,
        "uint32": gdal.GDT_UInt32,
        "int32": gdal.GDT_Int32,
        "float32": gdal.GDT_Float32,
        "float64": gdal.GDT_Float64,
    }
    
    datatype = dtype_to_gdal.get(im_data.dtype.name, gdal.GDT_Float32)
    
    # Get dimensions
    if len(im_data.shape) == 3:
        im_bands, im_height, im_width = im_data.shape
    else:
        im_bands, (im_height, im_width) = 1, im_data.shape
    
    # Create directory if needed
    pathlib.Path(filename).parent.mkdir(parents=True, exist_ok=True)
    
    # Create dataset
    driver = gdal.GetDriverByName("GTiff")
    dataset = driver.Create(filename, im_width, im_height, im_bands, datatype)
    dataset.SetGeoTransform(im_geotrans)
    dataset.SetProjection(im_proj)
    
    # Write data
    if im_bands == 1:
        dataset.GetRasterBand(1).WriteArray(im_data)
    else:
        for i in range(im_bands):
            dataset.GetRasterBand(i + 1).WriteArray(im_data[i])
    
    dataset = None


def write_tiff(lonn: np.ndarray, latt: np.ndarray, savename: str, data: np.ndarray) -> None:
    """Write GeoTIFF file with WGS84 projection.
    
    Args:
        lonn: Longitude array.
        latt: Latitude array.
        savename: Output file path.
        data: Data array to save.
    """
    driver = gdal.GetDriverByName("GTiff")
    dataset = driver.Create(savename, len(lonn), len(latt), 1, gdal.GDT_Float32)
    
    spatial_ref = osr.SpatialReference()
    spatial_ref.ImportFromEPSG(4326)
    dataset.SetProjection(spatial_ref.ExportToWkt())
    dataset.SetGeoTransform(
        (lonn[0], lonn[1] - lonn[0], 0, latt[-1], 0, latt[-2] - latt[-1])
    )
    dataset.GetRasterBand(1).WriteArray(data)
    dataset = None

