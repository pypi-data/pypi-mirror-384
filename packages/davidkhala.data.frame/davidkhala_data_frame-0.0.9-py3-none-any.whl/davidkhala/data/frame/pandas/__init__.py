import pandas as pd
def upsert(df, *primary_keys: str, record: dict):
    index_keys = df.index.names if isinstance(df.index, pd.MultiIndex) else [df.index.name]
    condition = True

    for key in primary_keys:
        if key in df.columns:
            condition &= (df[key] == record[key])
        elif key in index_keys:
            condition &= (df.index.get_level_values(key) == record[key])
        else:
            raise KeyError(f"'{key}' not found in either columns or index")

    match_indices = df[condition].index

    if not match_indices.empty:
        for col, value in record.items():
            if col in df.columns:
                df.loc[match_indices, col] = value
    else:
        df = pd.concat([df, pd.DataFrame([record])], ignore_index=True)

    return df