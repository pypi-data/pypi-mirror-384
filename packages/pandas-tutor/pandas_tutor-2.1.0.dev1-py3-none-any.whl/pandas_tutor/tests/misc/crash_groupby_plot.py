import pandas as pd
import io

csv = '''
breed,type,longevity,size,weight
German Shepherd,herding,9.73,large,
Beagle,hound,12.3,small,
Yorkshire Terrier,toy,12.6,small,5.5
Golden Retriever,sporting,12.04,medium,60.0
Bulldog,non-sporting,6.29,medium,45.0
Labrador Retriever,sporting,12.04,medium,67.5
Boxer,working,8.81,medium,
Poodle,non-sporting,11.95,medium,
Dachshund,hound,12.63,small,24.0
Rottweiler,working,9.11,large,
Boston Terrier,non-sporting,10.92,medium,
Shih Tzu,toy,13.2,small,12.5
Miniature Schnauzer,terrier,11.81,small,15.5
Doberman Pinscher,working,10.33,large,
Chihuahua,toy,16.5,small,5.5
Siberian Husky,working,12.58,medium,47.5
Pomeranian,toy,9.67,small,5.0
French Bulldog,non-sporting,9.0,medium,27.0
Great Dane,working,6.96,large,
Shetland Sheepdog,herding,12.53,small,22.0
Cavalier King Charles Spaniel,toy,11.29,small,15.5
German Shorthaired Pointer,sporting,11.46,large,62.5
Maltese,toy,12.25,small,5.0
'''

dogs = pd.read_csv(io.StringIO(csv))

(dogs
 .groupby('type').plot() # <-- crashes when i try to do a plot() here
)
