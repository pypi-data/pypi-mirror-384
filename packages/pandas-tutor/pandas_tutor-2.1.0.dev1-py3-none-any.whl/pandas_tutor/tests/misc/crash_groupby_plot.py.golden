{
  "code": "import pandas as pd\nimport io\n\ncsv = '''\nbreed,type,longevity,size,weight\nGerman Shepherd,herding,9.73,large,\nBeagle,hound,12.3,small,\nYorkshire Terrier,toy,12.6,small,5.5\nGolden Retriever,sporting,12.04,medium,60.0\nBulldog,non-sporting,6.29,medium,45.0\nLabrador Retriever,sporting,12.04,medium,67.5\nBoxer,working,8.81,medium,\nPoodle,non-sporting,11.95,medium,\nDachshund,hound,12.63,small,24.0\nRottweiler,working,9.11,large,\nBoston Terrier,non-sporting,10.92,medium,\nShih Tzu,toy,13.2,small,12.5\nMiniature Schnauzer,terrier,11.81,small,15.5\nDoberman Pinscher,working,10.33,large,\nChihuahua,toy,16.5,small,5.5\nSiberian Husky,working,12.58,medium,47.5\nPomeranian,toy,9.67,small,5.0\nFrench Bulldog,non-sporting,9.0,medium,27.0\nGreat Dane,working,6.96,large,\nShetland Sheepdog,herding,12.53,small,22.0\nCavalier King Charles Spaniel,toy,11.29,small,15.5\nGerman Shorthaired Pointer,sporting,11.46,large,62.5\nMaltese,toy,12.25,small,5.0\n'''\n\ndogs = pd.read_csv(io.StringIO(csv))\n\n(dogs\n .groupby('type').plot() # <-- crashes when i try to do a plot() here\n)\n",
  "explanation": [
    {
      "type": "GroupByCall",
      "code_step": "(dogs\n .groupby('type').plot() # <-- crashes when i try to do a plot() here\n)",
      "fragment": {
        "start": {
          "line": 1,
          "ch": 1
        },
        "end": {
          "line": 1,
          "ch": 17
        }
      },
      "mapping": [
        {
          "illustrate": "highlight",
          "select": "column",
          "anchor": "lhs",
          "label": "type"
        }
      ],
      "data_frame": {
        "lhs": {
          "type": "DataFrame",
          "col_names": [
            "breed",
            "type",
            "longevity",
            "size",
            "weight"
          ],
          "row_labels": [
            0,
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            9,
            10,
            11,
            12,
            13,
            14,
            15,
            16,
            17,
            18,
            19,
            20,
            21,
            22
          ],
          "data": [
            [
              "German Shepherd",
              "herding",
              9.73,
              "large",
              null
            ],
            [
              "Beagle",
              "hound",
              12.3,
              "small",
              null
            ],
            [
              "Yorkshire Terrier",
              "toy",
              12.6,
              "small",
              5.5
            ],
            [
              "Golden Retriever",
              "sporting",
              12.04,
              "medium",
              60.0
            ],
            [
              "Bulldog",
              "non-sporting",
              6.29,
              "medium",
              45.0
            ],
            [
              "Labrador Retriever",
              "sporting",
              12.04,
              "medium",
              67.5
            ],
            [
              "Boxer",
              "working",
              8.81,
              "medium",
              null
            ],
            [
              "Poodle",
              "non-sporting",
              11.95,
              "medium",
              null
            ],
            [
              "Dachshund",
              "hound",
              12.63,
              "small",
              24.0
            ],
            [
              "Rottweiler",
              "working",
              9.11,
              "large",
              null
            ],
            [
              "Boston Terrier",
              "non-sporting",
              10.92,
              "medium",
              null
            ],
            [
              "Shih Tzu",
              "toy",
              13.2,
              "small",
              12.5
            ],
            [
              "Miniature Schnauzer",
              "terrier",
              11.81,
              "small",
              15.5
            ],
            [
              "Doberman Pinscher",
              "working",
              10.33,
              "large",
              null
            ],
            [
              "Chihuahua",
              "toy",
              16.5,
              "small",
              5.5
            ],
            [
              "Siberian Husky",
              "working",
              12.58,
              "medium",
              47.5
            ],
            [
              "Pomeranian",
              "toy",
              9.67,
              "small",
              5.0
            ],
            [
              "French Bulldog",
              "non-sporting",
              9.0,
              "medium",
              27.0
            ],
            [
              "Great Dane",
              "working",
              6.96,
              "large",
              null
            ],
            [
              "Shetland Sheepdog",
              "herding",
              12.53,
              "small",
              22.0
            ],
            [
              "Cavalier King Charles Spaniel",
              "toy",
              11.29,
              "small",
              15.5
            ],
            [
              "German Shorthaired Pointer",
              "sporting",
              11.46,
              "large",
              62.5
            ],
            [
              "Maltese",
              "toy",
              12.25,
              "small",
              5.0
            ]
          ]
        },
        "rhs": {
          "type": "DataFrameGroupBy",
          "col_names": [
            "breed",
            "type",
            "longevity",
            "size",
            "weight"
          ],
          "row_labels": [
            0,
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            9,
            10,
            11,
            12,
            13,
            14,
            15,
            16,
            17,
            18,
            19,
            20,
            21,
            22
          ],
          "data": [
            [
              "German Shepherd",
              "herding",
              9.73,
              "large",
              null
            ],
            [
              "Beagle",
              "hound",
              12.3,
              "small",
              null
            ],
            [
              "Yorkshire Terrier",
              "toy",
              12.6,
              "small",
              5.5
            ],
            [
              "Golden Retriever",
              "sporting",
              12.04,
              "medium",
              60.0
            ],
            [
              "Bulldog",
              "non-sporting",
              6.29,
              "medium",
              45.0
            ],
            [
              "Labrador Retriever",
              "sporting",
              12.04,
              "medium",
              67.5
            ],
            [
              "Boxer",
              "working",
              8.81,
              "medium",
              null
            ],
            [
              "Poodle",
              "non-sporting",
              11.95,
              "medium",
              null
            ],
            [
              "Dachshund",
              "hound",
              12.63,
              "small",
              24.0
            ],
            [
              "Rottweiler",
              "working",
              9.11,
              "large",
              null
            ],
            [
              "Boston Terrier",
              "non-sporting",
              10.92,
              "medium",
              null
            ],
            [
              "Shih Tzu",
              "toy",
              13.2,
              "small",
              12.5
            ],
            [
              "Miniature Schnauzer",
              "terrier",
              11.81,
              "small",
              15.5
            ],
            [
              "Doberman Pinscher",
              "working",
              10.33,
              "large",
              null
            ],
            [
              "Chihuahua",
              "toy",
              16.5,
              "small",
              5.5
            ],
            [
              "Siberian Husky",
              "working",
              12.58,
              "medium",
              47.5
            ],
            [
              "Pomeranian",
              "toy",
              9.67,
              "small",
              5.0
            ],
            [
              "French Bulldog",
              "non-sporting",
              9.0,
              "medium",
              27.0
            ],
            [
              "Great Dane",
              "working",
              6.96,
              "large",
              null
            ],
            [
              "Shetland Sheepdog",
              "herding",
              12.53,
              "small",
              22.0
            ],
            [
              "Cavalier King Charles Spaniel",
              "toy",
              11.29,
              "small",
              15.5
            ],
            [
              "German Shorthaired Pointer",
              "sporting",
              11.46,
              "large",
              62.5
            ],
            [
              "Maltese",
              "toy",
              12.25,
              "small",
              5.0
            ]
          ],
          "group_data": {
            "col_names": [
              "type"
            ],
            "groups": [
              {
                "name": [
                  "herding"
                ],
                "labels": [
                  0,
                  19
                ]
              },
              {
                "name": [
                  "hound"
                ],
                "labels": [
                  1,
                  8
                ]
              },
              {
                "name": [
                  "non-sporting"
                ],
                "labels": [
                  4,
                  7,
                  10,
                  17
                ]
              },
              {
                "name": [
                  "sporting"
                ],
                "labels": [
                  3,
                  5,
                  21
                ]
              },
              {
                "name": [
                  "terrier"
                ],
                "labels": [
                  12
                ]
              },
              {
                "name": [
                  "toy"
                ],
                "labels": [
                  2,
                  11,
                  14,
                  16,
                  20,
                  22
                ]
              },
              {
                "name": [
                  "working"
                ],
                "labels": [
                  6,
                  9,
                  13,
                  15,
                  18
                ]
              }
            ]
          }
        }
      }
    },
    {
      "type": "AggCall",
      "code_step": "(dogs\n .groupby('type').plot() # <-- crashes when i try to do a plot() here\n)",
      "fragment": {
        "start": {
          "line": 1,
          "ch": 17
        },
        "end": {
          "line": 1,
          "ch": 24
        }
      },
      "mapping": [
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 0
          },
          "to": {
            "anchor": "rhs",
            "label": "herding"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 19
          },
          "to": {
            "anchor": "rhs",
            "label": "herding"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 1
          },
          "to": {
            "anchor": "rhs",
            "label": "hound"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 8
          },
          "to": {
            "anchor": "rhs",
            "label": "hound"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 4
          },
          "to": {
            "anchor": "rhs",
            "label": "non-sporting"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 7
          },
          "to": {
            "anchor": "rhs",
            "label": "non-sporting"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 10
          },
          "to": {
            "anchor": "rhs",
            "label": "non-sporting"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 17
          },
          "to": {
            "anchor": "rhs",
            "label": "non-sporting"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 3
          },
          "to": {
            "anchor": "rhs",
            "label": "sporting"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 5
          },
          "to": {
            "anchor": "rhs",
            "label": "sporting"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 21
          },
          "to": {
            "anchor": "rhs",
            "label": "sporting"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 12
          },
          "to": {
            "anchor": "rhs",
            "label": "terrier"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 2
          },
          "to": {
            "anchor": "rhs",
            "label": "toy"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 11
          },
          "to": {
            "anchor": "rhs",
            "label": "toy"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 14
          },
          "to": {
            "anchor": "rhs",
            "label": "toy"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 16
          },
          "to": {
            "anchor": "rhs",
            "label": "toy"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 20
          },
          "to": {
            "anchor": "rhs",
            "label": "toy"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 22
          },
          "to": {
            "anchor": "rhs",
            "label": "toy"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 6
          },
          "to": {
            "anchor": "rhs",
            "label": "working"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 9
          },
          "to": {
            "anchor": "rhs",
            "label": "working"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 13
          },
          "to": {
            "anchor": "rhs",
            "label": "working"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 15
          },
          "to": {
            "anchor": "rhs",
            "label": "working"
          }
        },
        {
          "illustrate": "outline",
          "select": "row",
          "from": {
            "anchor": "lhs",
            "label": 18
          },
          "to": {
            "anchor": "rhs",
            "label": "working"
          }
        }
      ],
      "data_frame": {
        "lhs": "prev_rhs",
        "rhs": {
          "type": "Series",
          "row_labels": [
            "herding",
            "hound",
            "non-sporting",
            "sporting",
            "terrier",
            "toy",
            "working"
          ],
          "data": [
            "AxesSubplot(0.125,0.11;0.775x0.77)",
            "AxesSubplot(0.125,0.11;0.775x0.77)",
            "AxesSubplot(0.125,0.11;0.775x0.77)",
            "AxesSubplot(0.125,0.11;0.775x0.77)",
            "AxesSubplot(0.125,0.11;0.775x0.77)",
            "AxesSubplot(0.125,0.11;0.775x0.77)",
            "AxesSubplot(0.125,0.11;0.775x0.77)"
          ]
        }
      }
    }
  ]
}